File: 01 - Arrays/25 - majority_ele_2.cpp
vector<int> majorityElementII(vector<int> &arr)
{
    //    Write your code here
    int ct1=0, ct2=0, candidate1=0, candidate2=0;
    int length = arr.size();
    for(int i=0; i<length; i++)
    {
        if(arr[i]==candidate1)
            ct1++;
        else if(arr[i]==candidate2)
            ct2++;
        else if(ct1==0)
        {
            candidate1 = arr[i];
            ct1 = 1;
        }
        else if(ct2==0)
        {
            candidate2 = arr[i];
            ct2 = 1;
        }
        else
        {
            ct1--;
            ct2--;
        }
    }
    
    
    int count_c1=0, count_c2=0;
    for(auto& it: arr)
    {
        if(it==candidate1)
            count_c1++;
        else if(it==candidate2)
            count_c2++;
    }
    vector <int> ans;
    if(count_c1 > length/3)
        ans.push_back(candidate1);
    if(count_c2 > length/3)
        ans.push_back(candidate2);
    return ans; 
}

File: 01 - Arrays/25 - majority_ele_2.py
# https://leetcode.com/problems/majority-element-ii/discuss/63520/Boyer-Moore-Majority-Vote-algorithm-and-my-elaboration
# https://www.codingninjas.com/codestudio/problem-details/majority-element-ii_893027

    def majorityElementII(arr):
        # Write your code here.
        if not arr:
            return []

        ct1, ct2, candidate1, candidate2 = 0, 0, 0, 0

        for ele in arr:

            if ele == candidate1:
                ct1 += 1
            elif ele == candidate2:
                ct2 += 1
            elif ct1 == 0:
                candidate1 = ele 
                ct1 = 1
            elif ct2 == 0:
                candidate2 = ele 
                ct2 = 1
            else:
                ct1, ct2 = ct1-1, ct2-1
        
        l = len(arr)//3
        return [n for n in (candidate1, candidate2) 
            if arr.count(n) > l]

        
tc = int(input())

for _ in range(tc):
    n = input()
    arr = [int(x) for x in input().split()]
    sol = majorityElementII(arr)
    print(*sol)